nautilus_core/portfolio/src/portfolio.rs

```diff
@@ -16,36 +16,141 @@
-    enums::OrderSide,
+    data::{quote::QuoteTick, Data},
+    enums::{OrderSide, OrderType, PositionSide, PriceType},
     events::{account::state::AccountState, order::OrderEventAny, position::PositionEvent},
     identifiers::{InstrumentId, Venue},
     instruments::any::InstrumentAny,
+    orders::any::OrderAny,
     position::Position,
-    types::money::Money,
+    types::{currency::Currency, money::Money, price::Price},
+};
+use rust_decimal::{
+    prelude::{FromPrimitive, ToPrimitive},
+    Decimal,
 };
-use rust_decimal::Decimal;
+use ustr::Ustr;
+
+use crate::manager::AccountsManager;
+
+struct UpdateQuoteTickHandler {
+    id: Ustr,
+    callback: Box<dyn Fn(&QuoteTick)>,
+}
+
+impl MessageHandler for UpdateQuoteTickHandler {
+    fn id(&self) -> Ustr {
+        self.id
+    }
+
+    fn handle(&self, msg: &dyn Any) {
+        (self.callback)(msg.downcast_ref::<&QuoteTick>().unwrap());
+    }
+    fn handle_response(&self, _resp: DataResponse) {}
+    fn handle_data(&self, _data: Data) {}
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
+
+struct UpdateOrderHandler {
+    id: Ustr,
+    callback: Box<dyn Fn(&OrderEventAny)>,
+}
+
+impl MessageHandler for UpdateOrderHandler {
+    fn id(&self) -> Ustr {
+        self.id
+    }
+
+    fn handle(&self, msg: &dyn Any) {
+        (self.callback)(msg.downcast_ref::<&OrderEventAny>().unwrap());
+    }
+    fn handle_response(&self, _resp: DataResponse) {}
+    fn handle_data(&self, _data: Data) {}
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
+
+struct UpdatePositionHandler {
+    id: Ustr,
+    callback: Box<dyn Fn(&PositionEvent)>,
+}
+
+impl MessageHandler for UpdatePositionHandler {
+    fn id(&self) -> Ustr {
+        self.id
+    }
+
+    fn handle(&self, msg: &dyn Any) {
+        (self.callback)(msg.downcast_ref::<&PositionEvent>().unwrap());
+    }
+    fn handle_response(&self, _resp: DataResponse) {}
+    fn handle_data(&self, _data: Data) {}
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
+
+struct UpdateAccountHandler {
+    id: Ustr,
+    callback: Box<dyn Fn(&AccountState)>,
+}
+
+impl MessageHandler for UpdateAccountHandler {
+    fn id(&self) -> Ustr {
+        self.id
+    }
+
+    fn handle(&self, msg: &dyn Any) {
+        (self.callback)(msg.downcast_ref::<&AccountState>().unwrap());
+    }
+    fn handle_response(&self, _resp: DataResponse) {}
+    fn handle_data(&self, _data: Data) {}
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
 
 pub struct Portfolio {
     clock: Rc<RefCell<dyn Clock>>,
     cache: Rc<RefCell<Cache>>,
     msgbus: Rc<RefCell<MessageBus>>,
-    accounts: HashMap<Venue, AccountAny>,
+    accounts: AccountsManager,
     analyzer: PortfolioAnalyzer,
-    // venue: Option<Venue>, // Added for completeness but was meant to be a "temporary hack"
     unrealized_pnls: HashMap<InstrumentId, Money>,
     realized_pnls: HashMap<InstrumentId, Money>,
     net_positions: HashMap<InstrumentId, Decimal>,
```

Comment:
Perhaps for portfolio we can take a slightly different approach to inner outer. These lines are the actual inner state of the portfolio.  Perhaps they can be wrapped into their own struct, and rc ref celled on this Portfolio struct (which will act as outer).

```rust
struct PorfolioState {
    accounts: AccountsManager,
    analyzer: PortfolioAnalyzer,
    unrealized_pnls: HashMap<InstrumentId, Money>,
    realized_pnls: HashMap<InstrumentId, Money>,
    net_positions: HashMap<InstrumentId, Decimal>,
}

struct Porfolio {
  ...
  inner. Rc<RefCell<PortfolioState>>
```

This way you can clone the portfolio state and pass it to the closure and relevant functions without needing to wrap the whole portfolio struct which might be cumbersome. However, it will mean that all the functions that the closures call will have to be changed slightly to receive the portfolio state, message bus, cache as separate parameter.

```
update_quote_tick(&self, ,.....)
update_qutote_tick(cache, messagebus, portfolio_state, ....)

---

To abstract away the borrow and borrow mut calls. You can also consider an additional thin helper layer.

```rust
struct PorfolioWrapper {
   inner: Rc<RefCell<PortfolioState>>,
}

struct Porfolio {
    inner: PortfolioWrapper
 }
```

For PortfolioWrapper you can consider implementing something like these helper methods to easily access the inner fields like

```rust
impl PortfolioWrapper {
    fn accounts(&self) -> impl Deref<Target = &AccountsManager> {
        self.inner.borrow().accounts
    }
}
```